File: web/src/lib/billsnapshot.test.js
// snapshot round-trip parity -- same shape/semantics as android BillSnapshotTest.kt
import { describe, it, expect } from 'vitest'
import { snapshotBill, toBillTrips, tourSortKey, purposeLineForTrip, institutesForTrip, buildStoredTrips } from './billsnapshot.js'
import { billTotals, makeTrip, leg } from './billmath.js'

// stored-shape trips (snake_case legs) as Bills.svelte buildSnapshotTrips produces them
const trips = [
  { tripId: 't1', purposeLine: 'Capacity Assessment - BACCO (Ref: X, 06 May 2026)', nights: 0, foodDays: 0,
    legs: [
      { dep_date: '2026-05-11', dep_time: '10:30', dep_place: 'A', arr_date: '2026-05-11', arr_time: '11:00', arr_place: 'B', mode: 'Uber Car', class: 'Rented', fare: 441, remarks: null },
      { dep_date: '2026-05-11', dep_time: '16:11', dep_place: 'B', arr_date: '2026-05-11', arr_time: '17:12', arr_place: 'A', mode: 'Bus', class: 'AC', fare: 482, remarks: null },
    ] },
  { tripId: 't2', purposeLine: 'ToT - JUTTI (Ref: Y, 04 Mar 2026)', nights: 6, foodDays: 6.5,
    legs: [
      { dep_date: '2026-06-08', dep_time: '14:54', dep_place: 'A', arr_date: '2026-06-08', arr_time: '15:31', arr_place: 'B', mode: 'Uber Bike', class: 'Rented', fare: 176, remarks: null },
      { dep_date: '2026-06-14', dep_time: '15:30', dep_place: 'B', arr_date: '2026-06-14', arr_time: '23:30', arr_place: 'A', mode: 'Launch', class: 'Cabin', fare: 1500, remarks: null },
    ] },
]

const totals = billTotals(trips.map((t) => makeTrip(t.legs.map((l) => leg(l.fare, l.dep_date)), '', '', t.nights, t.foodDays)))

it('real-sample totals survive the snapshot', () => {
  const snap = snapshotBill('2026-06-15', 'Test Officer', trips, totals)
  expect(snap.totals.net).toBeCloseTo(24349, 3)
  expect(snap.totals.words).toBe('Twenty Four Thousand Three Hundred Forty Nine Only')
  expect(snap.trips[0].legs[0].dep_date).toBe('2026-05-11') // stored shape kept intact
})

it('toBillTrips recomputes per-leg night/food deterministically', () => {
  const snap = snapshotBill('2026-06-15', 'Test Officer', trips, totals)
  const rendered = toBillTrips(snap)
  // trip1 claimed 0/0 -> whole trip zeroed
  expect(rendered[0].legs.map((l) => [l.nightStay, l.foodDay])).toEqual([[0, 0], [0, 0]])
  // trip2: first day full, last travel day half food / no night
  expect(rendered[1].legs.map((l) => [l.nightStay, l.foodDay])).toEqual([[1, 1], [0, 0.5]])
  expect(rendered[1].purposeLine).toBe('ToT - JUTTI (Ref: Y, 04 Mar 2026)')
})

// -- android T5 tour-grouping rules ported for web (chrono sort, multi-institute, purpose join) --

describe('tourSortKey', () => {
  it('uses the earliest leg departure when legs exist', () => {
    const legs = [
      { dep_date: '2026-07-05', dep_time: '14:00' },
      { dep_date: '2026-07-05', dep_time: '09:00' },
      { dep_date: '2026-07-06', dep_time: '08:00' },
    ]
    expect(tourSortKey(legs, '2026-07-01T00:00:00Z')).toBe('2026-07-05T09:00')
  })

  it('falls back to trip started_at with no legs', () => {
    expect(tourSortKey([], '2026-07-01T00:00:00Z')).toBe('2026-07-01T00:00:00Z')
  })
})

const pv = { trip_id: 't1', is_additional: false, institute: 'SDC Overseas Training and Testing Center, Dhaka', purpose: 'Capacity Assessment', association: 'BACCO', ref_no: 'X', ref_date: '2026-07-01', start_date: '2026-07-05' }
const addl1 = { trip_id: 't1', is_additional: true, institute: 'Lalmatia Mohila College, Dhaka', purpose: 'Monitoring Visit', association: 'BGMEA', ref_no: 'Y', ref_date: '2026-07-02', start_date: '2026-07-05' }
const addl2 = { trip_id: 't1', is_additional: true, institute: 'Kamarpara TTC, Dhaka', purpose: 'Capacity Assessment', association: 'BACCO', ref_no: 'X', ref_date: '2026-07-01', start_date: '2026-07-05' } // identical to pv's purpose line

describe('institutesForTrip', () => {
  it('lists all institutes, primary first', () => {
    expect(institutesForTrip('t1', [addl1, addl2, pv])).toEqual([
      'SDC Overseas Training and Testing Center, Dhaka',
      'Lalmatia Mohila College, Dhaka',
      'Kamarpara TTC, Dhaka',
    ])
  })

  it('single-visit tour is just that institute', () => {
    expect(institutesForTrip('t1', [pv])).toEqual([pv.institute])
  })
})

describe('purposeLineForTrip', () => {
  it('joins distinct purpose+association+ref lines with "; ", primary first', () => {
    const line = purposeLineForTrip('t1', [addl1, pv])
    expect(line).toBe('Capacity Assessment - BACCO (Ref: X, 01 Jul 2026); Monitoring Visit - BGMEA (Ref: Y, 02 Jul 2026)')
  })

  it('collapses an additional visit identical to the primary into one band', () => {
    const line = purposeLineForTrip('t1', [pv, addl2])
    expect(line).toBe('Capacity Assessment - BACCO (Ref: X, 01 Jul 2026)')
  })

  it('single-visit tour prints one band, no join', () => {
    expect(purposeLineForTrip('t1', [pv])).toBe('Capacity Assessment - BACCO (Ref: X, 01 Jul 2026)')
  })
})

describe('buildStoredTrips', () => {
  it('keeps caller-given tripId order and joins multi-visit purpose bands', () => {
    const legs = [
      { trip_id: 't1', dep_date: '2026-07-05', dep_time: '09:00', dep_place: 'A', arr_date: '2026-07-05', arr_time: '10:00', arr_place: 'B', mode: 'Bus', class: 'AC', fare: 100, remarks: null },
    ]
    const [stored] = buildStoredTrips(['t1'], [pv, addl1], legs)
    expect(stored.tripId).toBe('t1')
    expect(stored.purposeLine).toContain('; ')
    expect(stored.legs.length).toBe(1)
  })
})

File: android/app/src/test/java/bd/sicip/qavisit/domain/BillSnapshotTest.kt
// round-trip snapshotBill()/renderableFromSnapshot() -- everything the JSON schema actually
// carries must survive the trip through JsonObject and back untouched. per-leg night/food
// display values are deliberately NOT part of that contract (see file header comment on
// domain/BillSnapshot.kt), so they're covered separately below.
package bd.sicip.qavisit.domain

import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import org.junit.Assert.assertEquals
import org.junit.Test

class BillSnapshotTest {
    private val leg1 = SnapshotLeg(
        depDate = "2026-06-08", depTime = "08:00:00", depPlace = "Dhaka",
        arrDate = "2026-06-08", arrTime = "14:00:00", arrPlace = "Bhola",
        mode = "Launch", travelClass = "1st", fare = 176.0, remarks = "remark 1",
    )
    private val leg2 = SnapshotLeg(
        depDate = "2026-06-14", depTime = "09:00:00", depPlace = "Bhola",
        arrDate = "2026-06-14", arrTime = "15:00:00", arrPlace = "Dhaka",
        mode = "Launch", travelClass = null, fare = 1500.0, remarks = null,
    )
    private val trip = SnapshotTrip(
        tripId = "trip-1",
        purposeLine = "Purpose: QA visit - BASIS (Ref: REF-1, 08 Jun 2026)",
        nights = 6,
        foodDays = 6.5,
        legs = listOf(leg1, leg2),
    )
    private val totals = BillTotals(ta = 1676.0, accommodation = 12000.0, food = 9750.0, net = 23426.0)

    @Test fun `snapshot round trips through json`() {
        val json = snapshotBill(billDate = "2026-06-15", officerName = "Jane Officer", trips = listOf(trip), totals = totals)
        val restored = renderableFromSnapshot(json)

        assertEquals(
            BillSnapshot(billDate = "2026-06-15", officerName = "Jane Officer", trips = listOf(trip), totals = totals),
            restored,
        )
    }

    @Test fun `words field derives from net amount`() {
        val json = snapshotBill(billDate = "2026-06-15", officerName = "Jane Officer", trips = listOf(trip), totals = totals)
        val words = json.getValue("totals").jsonObject.getValue("words").jsonPrimitive.content
        assertEquals("${amountInWords(totals.net.toLong())} Only", words)
    }

    @Test fun `toBillTrips recomputes per-leg night and food defaults`() {
        val snapshot = BillSnapshot(billDate = "2026-06-15", officerName = "Jane Officer", trips = listOf(trip), totals = totals)
        val billTrips = snapshot.toBillTrips()

        assertEquals(1, billTrips.size)
        val legs = billTrips[0].legs
        // first day of a 6-night, 6.5-food trip: full night/food; last day (trip end): halved food, no night.
        assertEquals(1, legs[0].nightStay)
        assertEquals(1.0, legs[0].foodDay, 0.001)
        assertEquals(0, legs[1].nightStay)
        assertEquals(0.5, legs[1].foodDay, 0.001)
    }

    @Test fun `toBillTrips zeroes every leg when the trip was claimed fully zeroed`() {
        val zeroedTrip = trip.copy(nights = 0, foodDays = 0.0)
        val snapshot = BillSnapshot(billDate = "2026-05-11", officerName = "Jane Officer", trips = listOf(zeroedTrip), totals = totals)
        val legs = snapshot.toBillTrips()[0].legs
        legs.forEach {
            assertEquals(0, it.nightStay)
            assertEquals(0.0, it.foodDay, 0.001)
        }
    }
}
